paddle/fluid/operators/matmul_op.cc

```diff
@@ -339,6 +609,14 @@ class MatMulOpMaker : public framework::OpProtoAndCheckerMaker {
         )DOC")
         .SetDefault(false);
     AddAttr<float>("alpha", "The scale of Out").SetDefault(1.0f);
+    AddAttr<std::vector<int>>("last_dim",
+                              "(vector<int>) "
+                              "the shape of inputs and output.")
+        .SetDefault(std::vector<int>{-1, -1, -1});
+    AddAttr<bool>("is_test",
+                  "(bool, default false) Set to true for inference only, false "
+                  "for training. Some layers may run faster when this is true.")
```

Comment:
Why matmul need `is_test` attribute?
Why add `last_dim` attribute?

Reply:
> Why matmul need `is_test` attribute?
**Add the "is_test" attribute for inference mode and don't influence other requirement.** 
> Why add `last_dim` attribute?
**To decrease the count of matmul operator's attributes, but it will result in that it is only for the special dimensions of reshape and transpose.**

Reply:
> Add the "is_test" attribute for inference mode and don't influence other requirement.

Matmul is a common and base op, and it should not have the difference between train and inference.

Reply:
@luotao1 Score (inference) and Forward part of Training sometimes have difference. "is_test" attribute is used to distinguish between them and it's widely used in many OPs. For example:
* Batch Norm: Use fixed mean/variance instead of computing on batch
* Softmax: skip epson for performance improvement
* sequence_pool: don't create index buffer for performance improvement

The optimization here for attention (i.e. remove transpose/reshape via enhanced MatMul) only need apply to Inference only at this time. So we add "is_test" to contain our code.

In case this optimization need be applied to training also, we can add backward part and remove this "is_test" in fwd.